File: ext/rubyOF/TrueTypeFont.cpp
#include "TrueTypeFont.h"


const ofUnicode::range ALL_UNICODE_RANGES[] = {
	ofUnicode::Space,
	ofUnicode::IdeographicSpace,
	ofUnicode::Latin,
	ofUnicode::Latin1Supplement,
	ofUnicode::Greek,
	ofUnicode::Cyrillic,
	ofUnicode::Arabic,
	ofUnicode::ArabicSupplement,
	ofUnicode::ArabicExtendedA,
	ofUnicode::Devanagari,
	ofUnicode::HangulJamo,
	ofUnicode::VedicExtensions,
	ofUnicode::LatinExtendedAdditional,
	ofUnicode::GreekExtended,
	ofUnicode::GeneralPunctuation,
	ofUnicode::SuperAndSubScripts,
	ofUnicode::CurrencySymbols,
	ofUnicode::LetterLikeSymbols,
	ofUnicode::NumberForms,
	ofUnicode::Arrows,
	ofUnicode::MathOperators,
	ofUnicode::MiscTechnical,
	ofUnicode::BoxDrawing,
	ofUnicode::BlockElement,
	ofUnicode::GeometricShapes,
	ofUnicode::MiscSymbols,
	ofUnicode::Dingbats,
	ofUnicode::Hiragana,
	ofUnicode::Katakana,
	ofUnicode::HangulCompatJamo,
	ofUnicode::KatakanaPhoneticExtensions,
	ofUnicode::CJKLettersAndMonths,
	ofUnicode::CJKUnified,
	ofUnicode::DevanagariExtended,
	ofUnicode::HangulExtendedA,
	ofUnicode::HangulSyllables,
	ofUnicode::HangulExtendedB,
	ofUnicode::AlphabeticPresentationForms,
	ofUnicode::ArabicPresFormsA,
	ofUnicode::ArabicPresFormsB,
	ofUnicode::KatakanaHalfAndFullwidthForms,
	ofUnicode::KanaSupplement,
	ofUnicode::RumiNumericalSymbols,
	ofUnicode::ArabicMath,
	ofUnicode::MiscSymbolsAndPictographs,
	ofUnicode::Emoticons,
	ofUnicode::TransportAndMap,
	ofUnicode::EnclosedCharacters,
	ofUnicode::Uncategorized,
	ofUnicode::AdditionalEmoticons,
	ofUnicode::AdditionalTransportAndMap,
	ofUnicode::OtherAdditionalSymbols
};


const std::initializer_list<ofUnicode::range> ALL_UNICODE_ALPHABETS[] = {
	ofAlphabet::Emoji,
	ofAlphabet::Japanese,
	ofAlphabet::Chinese,
	ofAlphabet::Korean,
	ofAlphabet::Arabic,
	ofAlphabet::Devanagari,
	ofAlphabet::Latin,
	ofAlphabet::Greek,
	ofAlphabet::Cyrillic
};




void ofTtfSettings_setFontName(ofTrueTypeFontSettings& settings, std::string name)
{
	settings.fontName = name;
}

std::string ofTtfSettings_getFontName(const ofTrueTypeFontSettings& settings){
	return settings.fontName.string();
}

void ofTtfSettings_setFontSize(ofTrueTypeFontSettings& settings, int size)
{
	settings.fontSize = size;
}

int ofTtfSettings_getFontSize(const ofTrueTypeFontSettings& settings){
	return settings.fontSize;
}

void ofTtfSettings_setAntialiased(ofTrueTypeFontSettings& settings, bool aa){
	settings.antialiased = aa;
}

bool ofTtfSettings_isAntialiased(const ofTrueTypeFontSettings& settings){
	return settings.antialiased;
}




// TODO: Implement this.
void ofTtfSettings_addRanges(ofTrueTypeFontSettings& settings, Rice::Object rb_range_list){
	// settings.addRanges();
}

void ofTtfSettings_addRange(ofTrueTypeFontSettings& settings, Rice::Object rb_range_index){
	
	Rice::Object tmp_obj = rb_range_index;
	int i = tmp_obj.is_nil() ? -1 : from_ruby<int>(tmp_obj);
	ofUnicode::range rng = ALL_UNICODE_RANGES[i];
	
	settings.addRange(rng);
}

// NOTE: This is custom (not in the main OpenFrameworks API)
void ofTtfSettings_addAlphabet(ofTrueTypeFontSettings& settings, Rice::Object rb_range_index){
	
	Rice::Object tmp_obj = rb_range_index;
	int i = tmp_obj.is_nil() ? -1 : from_ruby<int>(tmp_obj);
	
	std::initializer_list<ofUnicode::range> rngs = ALL_UNICODE_ALPHABETS[i];
	settings.addRanges(rngs);
}


using namespace Rice;

Rice::Class Init_rubyOF_trueTypeFont(Rice::Module rb_mRubyOF)
{
	Data_Type<ofTrueTypeFont> rb_cTrueTypeFont = 
		define_class_under<ofTrueTypeFont>(rb_mRubyOF, "TrueTypeFont");
	
	// typedef ofTrueTypeFont (ofTrueTypeFont::*constructTrueTypeFont)();
	
	rb_cTrueTypeFont
		.define_constructor(Constructor<ofTrueTypeFont>())
		.define_method("load",
			static_cast<bool (ofTrueTypeFont::*)
			(const ofTrueTypeFontSettings&)
			>(&ofTrueTypeFont::load)
		)
		.define_method("draw_string",      &ofTrueTypeFont::drawString)
		
		.define_method("string_width",     &ofTrueTypeFont::stringWidth)
		.define_method("string_height",    &ofTrueTypeFont::stringHeight)
		.define_method("string_bb",        &ofTrueTypeFont::getStringBoundingBox)
		// ^ need to bind the ofRectangle type before this will work correctly
		
		.define_method("size",             &ofTrueTypeFont::getSize)
		.define_method("line_height=",     &ofTrueTypeFont::setLineHeight)
		.define_method("line_height",      &ofTrueTypeFont::getLineHeight)
		.define_method("ascender_height",  &ofTrueTypeFont::getAscenderHeight)
		.define_method("descender_height", &ofTrueTypeFont::getDescenderHeight)
		.define_method("letter_spacing",   &ofTrueTypeFont::getLetterSpacing)
		.define_method("letter_spacing=",  &ofTrueTypeFont::setLetterSpacing)
		.define_method("space_size",       &ofTrueTypeFont::getSpaceSize)
		.define_method("space_size=",      &ofTrueTypeFont::setSpaceSize)
		// .define_method("getKerning",       &ofTrueTypeFont::getKerning)
			// compile error:
			// "error: ‘int ofTrueTypeFont::getKerning(int, int) const’ is protected"
		
		.define_method("antialiased?",     &ofTrueTypeFont::isAntiAliased)
		
		
		
		.define_method("get_string_mesh",  &ofTrueTypeFont::getStringMesh)
		// Returns the string as an ofMesh. Note: this is a mesh that contains vertices and texture coordinates for the textured font, not the points of the font that are returned via any of the get points functions.
		// src: OpenFrameworks documentation
		// 
		// This mesh is just bounding boxes for each and every character.
		
		.define_method("font_texture",     &ofTrueTypeFont::getFontTexture)
		// This is the texture associated with the mesh above
	;
	
	
	
	
	Data_Type<ofTrueTypeFontSettings> rb_cTrueTypeFontSettings = 
		define_class_under<ofTrueTypeFontSettings>(rb_mRubyOF, "TrueTypeFontSettings");
	
	rb_cTrueTypeFontSettings
		.define_constructor(Constructor<ofTrueTypeFontSettings, const std::string, int>())
		.define_method("font_name",       &ofTtfSettings_getFontName)
		.define_method("font_name=",      &ofTtfSettings_setFontName)
		.define_method("font_size",       &ofTtfSettings_getFontSize)
		.define_method("font_size=",      &ofTtfSettings_setFontSize)
		
		.define_method("antialiased?",    &ofTtfSettings_isAntialiased)
		.define_method("antialiased=",    &ofTtfSettings_setAntialiased)
		
		.define_method("add_ranges",      &ofTtfSettings_addRanges)
		.define_method("add_range",       &ofTtfSettings_addRange)
		
		.define_method("add_alphabet",    &ofTtfSettings_addAlphabet)
	;
	
	
	
	
	return rb_cTrueTypeFont;
}



 //    std::filesystem::path fontName;
	// int fontSize;
	// bool antialiased = true;
	// bool contours = false;
	// float simplifyAmt = 0.3f;
	// int dpi = 0;

	// enum Direction{
	// 	LeftToRight,
	// 	RightToLeft
	// };
	// Direction direction = LeftToRight;

	// void addRanges(std::initializer_list<ofUnicode::range> alphabet){
	// 	ranges.insert(ranges.end(), alphabet);
	// }

	// void addRange(const ofUnicode::range & range){
	// 	ranges.push_back(range);
	// }

File: lib/rubyOF/true_type_font.rb
# NOTE: SelfMonkeyPatch is a module that holds monkey patches that were created to define core behavior at the ruby level, instead of having to write C / C++.

module RubyOF
	module SelfMonkeyPatch


# This extension to the base font class allows you
# to read the name off the font object,
# without having to keep the settings object alive.
# All other properties of the font can be read
# through the font object, but not the name.
# It appears that the name is not even being set
# as a member variable on the font object.

# TODO: file bug report for ofTrueTypeFont, stating that name can not be retrieved from the font object, only the settings object.
module TrueTypeFont
	def load(settings)
		@settings = settings
		flag = super(settings)
		unless flag
			raise "ERROR: Could not load font #{@settings.font_name}"
		end
	end
	
	def name
		@settings.font_name
	end
	
	# Return the font name as specified by the settings object
	# (in some cases, this may actually be the full path to the font file)
	# 
	# (Currently, this is the indentifier of the font file, and not actually the 'name' of the font face.)
	def name
		raise "ERROR: Font not yet loaded." if @name.nil?
		# ^ @name not set until #load() is called.
		
		return @name
	end
	# NOTE: no setter for this value, because you can't change the font face once the font object is loaded. (have to switch objects)
end


end; end
# --- end monkey patches



module RubyOF

class TrueTypeFont
	# The actual definition is above, but we want to use clean monkey patching
	# (https://stackoverflow.com/questions/4470108/when-monkey-patching-a-method-can-you-call-the-overridden-method-from-the-new-i)
	
	prepend RubyOF::SelfMonkeyPatch::TrueTypeFont
	
	
	# # private :load_from_struct
	# def foo(data)
		
	# 	# convert symbol into integer position,
	# 	# and then convert position into the actual struct.
		
	# 	i = UnicodeRanges.index(:Latin)
	# 	struct = bar(i)
	# 	settings.addRange(struct)
		
	# 	self.load_from_struct(i)
	# end
	
	
	# Create easy gemfile-style DSL for loading font parameters,
	# including error checking.
	class << self
		def dsl_load # &block
			config = DSL_Object.new
			
			yield config
			
			
			font_settings = RubyOF::TrueTypeFontSettings.new(config.path, config.size)
			
			config.alphabets.each do |x|
				font_settings.add_alphabet x
			end
			
			
			
			puts "loading font"
			font = RubyOF::ResourceManager.instance.load font_settings
			# ^ will throw exception if font can not be loaded
			
			return font
		end
	end
	
	class DSL_Object
		attr_reader :alphabets
		attr_accessor :path, :size, :antialiased
		
		def initialize
			@alphabets = Array.new
		end
		
		def add_alphabet(alphabet)
			alphabet_list = RubyOF::TrueTypeFontSettings::UnicodeAlphabets
			unless alphabet_list.include? alphabet
				message = [
					"Alphabet '#{alphabet}' is not one of the supported values.",
					"Use a value from TrueTypeFontSettings::UnicodeAlphabets",
					"Try one of these: #{alphabet_list.inspect}"
				].join("\n")
				
				raise message
			end
			
			@alphabets << alphabet
		end
	end
end

class TrueTypeFontSettings
	UnicodeRanges = [
		:Space,
		:IdeographicSpace,
		:Latin,
		:Latin1Supplement,
		:Greek,
		:Cyrillic,
		:Arabic,
		:ArabicSupplement,
		:ArabicExtendedA,
		:Devanagari,
		:HangulJamo,
		:VedicExtensions,
		:LatinExtendedAdditional,
		:GreekExtended,
		:GeneralPunctuation,
		:SuperAndSubScripts,
		:CurrencySymbols,
		:LetterLikeSymbols,
		:NumberForms,
		:Arrows,
		:MathOperators,
		:MiscTechnical,
		:BoxDrawing,
		:BlockElement,
		:GeometricShapes,
		:MiscSymbols,
		:Dingbats,
		:Hiragana,
		:Katakana,
		:HangulCompatJamo,
		:KatakanaPhoneticExtensions,
		:CJKLettersAndMonths,
		:CJKUnified,
		:DevanagariExtended,
		:HangulExtendedA,
		:HangulSyllables,
		:HangulExtendedB,
		:AlphabeticPresentationForms,
		:ArabicPresFormsA,
		:ArabicPresFormsB,
		:KatakanaHalfAndFullwidthForms,
		:KanaSupplement,
		:RumiNumericalSymbols,
		:ArabicMath,
		:MiscSymbolsAndPictographs,
		:Emoticons,
		:TransportAndMap,
		:EnclosedCharacters,
		:Uncategorized,
		:AdditionalEmoticons,
		:AdditionalTransportAndMap,
		:OtherAdditionalSymbols
	]
	
	UnicodeAlphabets = [
		:Emoji,
		:Japanese,
		:Chinese,
		:Korean,
		:Arabic,
		:Devanagari,
		:Latin,
		:Greek,
		:Cyrillic
	]
	
	private :add_ranges
	private :add_range
	
	# Given a symbol from the UnicodeRanges list,
	# convert that to a numerical index.
	# 
	# On the C++ side, that number will be converted to the actual
	# range needed by the font system.
	def add_unicode_range(range_name)
		@ranges ||= Array.new
		@ranges << range_name
		i = UnicodeRanges.index(range_name)
		add_range(i)
	end
	
	
	alias :cpp_add_alphabet :add_alphabet
	def add_alphabet(alphabet_name)
		@alphabets ||= Array.new
		@alphabets << alphabet_name
		i = UnicodeAlphabets.index(alphabet_name)
		cpp_add_alphabet(i)
	end
	
	
	
	
	def ==(other)
		return true if other.equal?(self)
		return false unless other.instance_of?(self.class)
		
		
		return  (self.font_name == other.font_name &&
		         self.font_size == other.font_size &&
		         self.antialiased? == other.antialiased? &&
		         @ranges == other.instance_variable_get('@ranges') &&
		         @alphabets == other.instance_variable_get('@alphabets') )
	end
	
	
	
	# Need to implement both eql? and hash to use this class as Hash key
	# Resource manager needs to be able to use TrueTypeFontSettings as a key.
	def eql?(other)
		# by default this is the same as ==\
		
		# The eql? method returns true if obj and anObject have the same value. Used by Hash to test members for equality. For objects of class Object, eql? is synonymous with ==. Subclasses normally continue this tradition, but there are exceptions. Numeric types, for example, perform type conversion across ==, but not across eql?
		# src: http://ruby-doc.org/core-1.9.3/Object.html#method-i-eql-3F
		
		self == other
	end
	
	def hash
		[
			self.font_name,
			self.font_size,
			self.antialiased?,
			@ranges,
			@alphabets
		].hash
	end
end



end
